### PR TITLE
impl proper_controller_dispose DartFix

### PR DESCRIPTION
**Pull request by imsamgarg**
# Description

Dart Fix for proper_controller_dispose lint.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I have read the [CONTRIBUTING.md][contributing_link] document.
- [x] I have performed a self-review of my code.
- [ ] I have linked the issue ticket in the description.
- [ ] I have made the necessary changes to the documentation.
- [x] I have checked the formatting with `dart format .`
- [x] I have analyzed my code with `dart analyze .`
- [x] I have run `dart run custom_lint example` to check for any custom linting issues.

[contributing_link]: https://github.com/charlescyt/pyramid_lint/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by charlescyt** [comment]
Your implementation adds multiple dispose statements and I am not sure if this is the behaviour we want. Typically, quick fixes correspond to its own lint warning so when we apply a quick fix to a particular warning, it should only add one dispose statement.

When `custom_lint fix --apply` is available, I believe each fix will be applied to its corresponding lint warning and therefore the current behaviour may conflict with the command.

File: lib/src/lints/flutter/proper_controller_dispose.dart [hunk 1]
```diff
@@ -120,4 +90,136 @@ class ProperControllerDispose extends DartLintRule {
...
+
+        changeBuilder.addDartFileEdit((builder) {
+          builder.addSimpleInsertion(offset!, '''
+
+@override
+void dispose() {
+$disposeList
+super.dispose();
+} 
+ ''');
```
Comment (charlescyt) [inline lib/src/lints/flutter/proper_controller_dispose.dart:184]: The logic here is a bit hard to follow.

I would probably to something like the below and use `builder.addInsertion()` instead

```dart
// This is just a draft
final initStateMethod = node.members.findMethodDeclarationByName(
  'initState',
);

final buildMethod = node.members.findMethodDeclarationByName('build');

final (int offset, bool addNewLineAtTheStart, bool addNewLineAtTheEnd) =
    switch ((initStateMethod, buildMethod)) {
  (null, null) => (node.end, true, false),
  (null, final buildMethod?) => (buildMethod.offset, true, true),
  (final initStateMethod?, null) => (initStateMethod.end, true, false),
  (final _?, final buildMethod?) => (buildMethod.offset, false, true),
};

changeBuilder.addDartFileEdit((builder) {
  builder.addInsertion(offset, (builder) {
    if (addNewLineAtTheStart) builder.write('\n');
    builder.write('  @override\n');
    builder.write('  void dispose() {\n');
    for (final c in controllerDeclarations) {
      builder.write('    ${c.name.lexeme}.dispose();\n');
    }
    builder.write('    super.dispose();\n');
    builder.write('  }\n');
    if (addNewLineAtTheEnd) builder.write('\n');
  });
});
```

**Review by charlescyt — COMMENTED** [review]
I should probably rewrite the lint. Instead of traversing ClassDeclaration, we should probably start with FieldDeclaration.

**Comment by imsamgarg** [comment]
Changes done. Please review it.

File: lib/src/lints/flutter/proper_controller_dispose.dart [hunk 2]
```diff
@@ -31,91 +32,166 @@ class ProperControllerDispose extends DartLintRule {
...
+        final (int offset, bool addNewLineAtTheStart, bool addNewLineAtTheEnd) =
+            switch ((initStateMethod, buildMethod)) {
+          (null, null) => (node.end, true, false),
+          (null, final buildMethod?) => (buildMethod.offset, true, true),
+          (final initStateMethod?, null) => (initStateMethod.end, true, false),
+          (final _?, final buildMethod?) => (buildMethod.offset, false, true),
+        };
+
+        /// Always pick the first variable.
+        final variableName = controllerVariables.first.name;
```
Comment (charlescyt) [inline lib/src/lints/flutter/proper_controller_dispose.dart:153]: Hmm.. not sure why we always use the first one

File: lib/src/utils/type_checker.dart [hunk 3]
```diff
@@ -15,6 +15,11 @@ const statefulWidgetChecker = TypeChecker.fromName(
   packageName: 'flutter',
 );
 
+const stateChecker = TypeChecker.fromName(
+  'State',
+  packageName: 'flutter',
+);
```
[the remark this answers is not in the document]
Comment (imsamgarg) [reply lib/src/utils/type_checker.dart:21]: Damn!! I completely overlooked that.

File: lib/src/lints/flutter/proper_controller_dispose.dart (hunk 2 above)
Comment (imsamgarg) [reply lib/src/lints/flutter/proper_controller_dispose.dart:153]: I think the default analyzer's add 'require' param does the same sometimes, that's why I use the first one. We can check for the intersection point for the variable being the assist triggered upon. If you have any other ideas please comment.

File: lib/src/lints/flutter/proper_controller_dispose.dart [hunk 4]
```diff
@@ -31,91 +32,166 @@ class ProperControllerDispose extends DartLintRule {
...
-            return null;
+            return switch (e.expression) {
+              final MethodInvocation exp when exp.target is SimpleIdentifier =>
+                (exp.target! as SimpleIdentifier).name,
+              _ => null
+            };
           })
-          .whereType<String>()
+          .whereNotNull()
           .toList();
```
[the remark this answers is not in the document]
Comment (imsamgarg) [reply lib/src/lints/flutter/proper_controller_dispose.dart:87]: Yeah right, i completely forgot about that dispose method check.

**Comment by imsamgarg** [comment]
Changes done, please review the changes.

**Review by charlescyt — APPROVED** [review]
Looks good to me.

Although the formatting for adding the `State`'s dispose method is a bit off, it's not a big deal. I will see if I can do something about it.

Thanks for the PR!
